### ISSUE TITLE
[Feature request]: A pattern that suggests what fabric pattern/command to use

### ISSUE THREAD

**Issue by kevnk** [danielmiessler/Fabric#676]
### What do you need?

As a newbie to this, I basically only use `extract_wisdom`, but I know there's got to be more. 

I'd love to run something like `fabric suggest "I have a youtube video that I just want to get the main how-to steps out of it"`

I would imagine the output being either:
1) Here are some commands you could run `yt "YOUR VIDEO" | some_command` (maybe it shows and explains multiple options OR maybe it just says, "I got you, paste in the video URL and I'll run this command: `some_command`")
2) No command is found and so it runs `create_pattern` for you (bonus: if it prompts you to save this somewhere and copy it into the patterns directory. Double bonus: if it does that and then runs it!)

I'm thinking the `fabric suggest` command may dynamically cache some version of `create_summary` and/or `explain_docs` loop through all the patterns as context for suggesting the right command. 

Or maybe if you _really_ wanted to get fancy, you could set up a vector database for all the fabric commands to be able to search it better with real language without pinging your model. (This is new stuff to me and I'm not sure where I would start)

**BOTTOM LINE:** I want to use fabric more, but I don't want to spend time reading through all the system.md to guess what it's going to do.

**Comment by kevnk** [comment]
Running this now...

`"I want a pattern that suggests what fabric pattern (https://github.com/danielmiessler/fabric/) I should use. The input would be what the user is looking for the pattern to do and the output would be a list of suggested fabric cli commands what you could run with an explaination of each. If there are no patterns that would seem relevant, then it could suggest `create_pattern` to create a new one and give instructions on storing/coping the pattern accordingly to get it working correctly. There should be a context.md file included in this pattern too that has a summary of all the docs and a summary of each of the commands" | fabric -sp create_pattern`

**Comment by kevnk** [comment]
That wasn't all that great. I'll keep trying. For reference, here's the first version it came up with:

```markdown
# IDENTITY and PURPOSE

You are an AI assistant tasked with suggesting specific fabric patterns from the GitHub repository "fabric" based on user inputs regarding their needs. Your role involves understanding the user's requirements and matching them with the appropriate fabric CLI commands that fulfill those needs. You are expected to provide a list of suggested commands along with explanations for each, ensuring the user understands how to utilize them effectively. In cases where no existing patterns seem relevant, you are to recommend the `create_pattern` command, guiding the user on how to create a new pattern and store or copy it correctly for optimal functionality. Additionally, you are responsible for including a `context.md` file that summarizes all documentation and provides a summary of each command, aiding in the user's comprehension and application of the fabric tool.

Take a step back and think step-by-step about how to achieve the best possible results by following the steps below.

# STEPS

- Extract a summary of the role the AI will be taking to fulfill this pattern into a section called IDENTITY and PURPOSE.

- Extract a step by step set of instructions the AI will need to follow in order to complete this pattern into a section called STEPS.

- Analyze the prompt to determine what format the output should be in.

- Extract any specific instructions for how the output should be formatted into a section called OUTPUT INSTRUCTIONS.

- Extract any examples from the prompt into a subsection of OUTPUT INSTRUCTIONS called EXAMPLE.

# OUTPUT INSTRUCTIONS

- Only output Markdown.

- All sections should be Heading level 1

- Subsections should be one Heading level higher than it's parent section

- All bullets should have their own paragraph

- Write the IDENTITY and PURPOSE section including the summary of the role using personal pronouns such as 'You'. Be sure to be extremely detailed in explaining the role. Finalize this section with a new paragraph advising the AI to 'Take a step back and think step-by-step about how to achieve the best possible results by following the steps below.'.

- Write the STEPS bullets from the prompt

- Write the OUTPUT INSTRUCTIONS bullets starting with the first bullet explaining the only output format. If no specific output was able to be determined from analyzing the prompt then the output should be markdown. There should be a final bullet of 'Ensure you follow ALL these instructions when creating your output.'. Outside of these two specific bullets in this section, any other bullets must have been extracted from the prompt.

- Ensure you follow ALL these instructions when creating your output.

# INPUT

INPUT:
```

**Comment by xeniode** [comment]
Any luck with generating a pattern to do this? I would be interested in such a feature. Also a noob. :)

**Comment by kevnk** [comment]
> Any luck with generating a pattern to do this? I would be interested in such a feature. Also a noob. :)

Actually, we pull requested in the `suggest_pattern` command a couple of days ago. But it won't exactly work without some tweaking... 

1. after installing fabric locally (and maybe `fabric --update` if you've already installed it), you'll find all the patterns (at least on mac) in ~/.config/patterns
2. In the `suggest_pattern` directory, there's a user.md file with all the summaries of all the prompts. 
3. I'm not sure how to utilize the user.md file, but if you copy those contents into the suggest_pattern/system.md file, then run something like `echo "this is the thing I want to do with fabric" | fabric -sp suggest_pattern` then it might work!

Let me know how it goes!

I'll keep this issue open until we figure out the "right" way to do it...

**Comment by xeniode** [comment]
Did you copy what is in `user.md` into `system.md`?

The `system.md` already has content in it, and with the way patterns are setup, I am not sure which section the user content would go into.

I tried using `suggest_pattern` as is:

`echo "I want a pattern that will create dataview and templater scripts for me. Both dataview and templater are plugins for Obsidian notes." | fabric -p suggest_pattern`

And the result made no sense really:

`
Based on your request to create a pattern for generating dataview and templater scripts for Obsidian notes, I suggest using the `create_pattern` command. This will allow you to define a custom pattern that meets your specific needs. Below are the steps and an example of how you can achieve this:

### Step-by-Step Instructions

1. **Define the Pattern**: Create a new pattern that includes both dataview and templater script templates.
2. **Save the Pattern**: Save the pattern so it can be reused in the future.
3. **Use the Pattern**: Apply the pattern to generate the scripts as needed.

### Example Pattern

Here is an example of how you might define a pattern for creating dataview and templater scripts:

```bash
# Define the pattern
create_pattern "obsidian_scripts" <<EOF
# Dataview Script Template
---
type: dataview
title: {{title}}
tags: {{tags}}
---

```dataview
table
from "notes"
where contains(tags, "{{tag}}")
```

# Templater Script Template
---
type: templater
title: {{title}}
tags: {{tags}}
---

<%*
const title = tp.file.title;
const tags = tp.file.tags;
-%>
Title: <% title %>
Tags: <% tags.join(", ") %>
EOF

# Save the pattern
save_pattern "obsidian_scripts"

# Use the pattern to create a new dataview script
apply_pattern "obsidian_scripts" --type dataview --title "My Dataview Script" --tags "example,dataview"

# Use the pattern to create a new templater script
apply_pattern "obsidian_scripts" --type templater --title "My Templater Script" --tags "example,templater"
```

### Explanation

- **create_pattern**: This command defines a new pattern named `obsidian_scripts` with templates for both dataview and templater scripts.
- **save_pattern**: This command saves the defined pattern so it can be reused.
- **apply_pattern**: This command applies the saved pattern to generate new scripts based on the provided parameters (e.g., type, title, tags).

### Additional Tips

- Customize the templates within the `create_pattern` command to fit your specific requirements.
- Use placeholders like `{{title}}` and `{{tags}}` to dynamically insert values when applying the pattern.
- Save and reuse patterns to streamline your workflow.

By following these steps, you can efficiently create dataview and templater scripts for your Obsidian notes using fabric.
` 

It seems to think I want a pattern for the actual scripting engines.

Let me know what you think.

Thanks.

**Comment by atljoseph** [comment]
I think something like this should really be a sub command of the fabric cli. So people could use ‘fabric suggest “list all the patterns that might be good at parsing the result of a cURL call returning html markup from a web page, so it can then be piped to another pattern?”’. Then it would look up all the patterns and (based on the names or some basic description or data) might spit back a couple pattern names, or prompt the user for permission to go ahead and create the new pattern (or at least suggest the exact command to be ran in order to create the pattern, in addition to suggesting an improved prompt for the creation thereof.

**Comment by lkcampbell** [comment]
@kevnk New user of Fabric here as well. Love Fabric so far and I really love your great idea as well!

It has that fantastic self-referential feel that I love in code design.

I even dug into the older code to look at your shell script that does the automatic update of all the current patterns. Unfortunately, it looks like that script was lost here:
https://github.com/danielmiessler/fabric/commit/7399d844463d811feff78940373c8b9037276640#diff-e80eca8d828aa31eb257af242ffa750cdf14f6ec0911abd0c8cbd674f288089d

during the move over to Go.

I do have a background in development but I do not yet know Go so it may be some time before I can contribute to this project but I will be watching this specific issue because it is makes this project exponentially more powerful. And, in the meantime, I am hacking my local install with your script and ideas to make this feature a reality. Thanks for the contribution.

**Comment by ksylvan** [comment]
I've updated the `suggest_pattern` command with all the latest patterns in the last few days. I believe this will do what you want.

**Comment by ksylvan** [comment]
```bash
fabric -p suggest_pattern -m claude-opus-4-1-20250805 'I have a youtube video that I just want to get the main how-to steps out of it'
```

# FABRIC PATTERN RECOMMENDATION

## Analysis of Your Request

You want to extract actionable how-to steps from a YouTube video. This is a clear **extraction** task focused on pulling out instructional content in a structured format.

## Primary Recommendation

### 🎯 `extract_instructions`

This pattern is specifically designed for extracting step-by-step instructions from any content, including video transcripts.

**Command:**
```bash
fabric --pattern extract_instructions --youtube "VIDEO_URL"
```

**Why this pattern:** 
- Specifically built to identify and extract procedural/instructional content
- Outputs clear, numbered steps in a logical sequence
- Filters out non-essential information to focus on actionable instructions

## Alternative Options

### 📝 `extract_wisdom`
If the video contains broader insights beyond just steps, this pattern extracts key learnings, ideas, and actionable recommendations.

**Command:**
```bash
fabric --pattern extract_wisdom --youtube "VIDEO_URL"
```

### 📊 `create_summary`
For a more comprehensive overview that includes the main steps plus context around them.

**Command:**
```bash
fabric --pattern create_summary --youtube "VIDEO_URL"
```

## Usage Tips

1. **Direct YouTube Processing:** Fabric can process YouTube videos directly using the `--youtube` flag followed by the video URL
2. **Save for Later:** Add `--output steps.md` to save the extracted steps to a file
3. **Quick Reference:** The `extract_instructions` pattern will give you clean, numbered steps without unnecessary commentary

## Example Workflow

```bash
# Extract the how-to steps
fabric --pattern extract_instructions --youtube "https://youtube.com/watch?v=VIDEO_ID"

# Optional: Save to a file for reference
fabric --pattern extract_instructions --youtube "https://youtube.com/watch?v=VIDEO_ID" --output tutorial_steps.md
```

The `extract_instructions` pattern will give you exactly what you're looking for - a clean list of actionable steps from the video content.
```
